File: Problems/0800_0899/0855_Exam_Room/Project_Java/ExamRoom.java
import java.util.*;

class ExamRoom {
    // 467ms - 1282ms
    int N;
    ArrayList<Integer> indexs = new ArrayList<>();

    public ExamRoom(int n) {
        N = n;
    }

    public int seat() {
        if (indexs.size() == 0) {
            indexs.add(0);
            return 0;
        }
        int d = Math.max(indexs.get(0), N - 1 - indexs.get(indexs.size() - 1));
        for (int i = 0; i < indexs.size() - 1; i++) {
            d = Math.max(d, (indexs.get(i + 1) - indexs.get(i)) / 2);
        }
        if (indexs.get(0) == d) {
            indexs.add(0, 0);
            return 0;
        }
        for (int i = 0; i < indexs.size() - 1; i++) {
            if ((indexs.get(i + 1) - indexs.get(i)) / 2 == d) {
                indexs.add(i + 1, (indexs.get(i + 1) + indexs.get(i)) / 2);
                return indexs.get(i + 1);
            }
        }
        indexs.add(N - 1);
        return N - 1;
    }

    public void leave(int p) {
        for (int i = 0; i < indexs.size(); i++) {
            if (indexs.get(i) == p) {
                indexs.remove(i);
            }
        }
    }
}

/**
 * Your ExamRoom object will be instantiated and called as such:
 * ExamRoom obj = new ExamRoom(n);
 * int param_1 = obj.seat();
 * obj.leave(p);
 */

File: Problems/0800_0899/0855_Exam_Room/Project_Go/solution/ExamRoom.go
package solution

// 139ms - 143ms

type ExamRoom struct {
	N      int
	indexs []int
}

func Constructor(N int) ExamRoom {
	return ExamRoom{N, make([]int, 0)}
}

func (this *ExamRoom) Seat() int {
	if len((*this).indexs) == 0 {
		(*this).indexs = append((*this).indexs, 0)
		return 0
	}
	d := myMax((*this).indexs[0], (*this).N-1-(*this).indexs[len((*this).indexs)-1])
	for i := 0; i < len((*this).indexs)-1; i++ {
		d = myMax(d, ((*this).indexs[i+1]-(*this).indexs[i])/2)
	}
	if (*this).indexs[0] == d {
		(*this).indexs = append([]int{0}, (*this).indexs...)
		return 0
	}
	for i := 0; i < len((*this).indexs)-1; i++ {
		if ((*this).indexs[i+1]-(*this).indexs[i])/2 == d {
			(*this).indexs = append((*this).indexs[:i+1], append([]int{((*this).indexs[i+1] + (*this).indexs[i]) / 2}, (*this).indexs[i+1:]...)...)
			return (*this).indexs[i+1]
		}
	}
	(*this).indexs = append((*this).indexs, (*this).N-1)
	return (*this).N - 1
}

func (this *ExamRoom) Leave(p int) {
	for i := 0; i < len((*this).indexs); i++ {
		if (*this).indexs[i] == p {
			(*this).indexs = append((*this).indexs[:i], (*this).indexs[i+1:]...)
		}
	}
}

func myMax(a, b int) int {
	if a > b {
		return a
	}
	return b
}

/**
 * Your ExamRoom object will be instantiated and called as such:
 * obj := Constructor(N);
 * param_1 := obj.Seat();
 * obj.Leave(p);
 */
